testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -378,89 +473,223 @@ impl RelayerThread {
-                        "Relayer: Winner of sortition {} did not commit to the correct parent tenure. Attempt to continue tenure.",
-                        &sn.consensus_hash
-                    );
-                    // We didn't win the sortition, but the miner that did win
-                    // did not commit to the correct parent tenure. This means
-                    // it will be unable to produce a valid block, so we should
-                    // continue our tenure.
-                    MinerDirective::ContinueTenure {
-                        new_burn_view: sn.consensus_hash,
-                    }
-                } else {
-                    info!("Relayer: Stop tenure");
-                    MinerDirective::StopTenure
-                },
-            )
-        } else {
-            // find out what epoch the Stacks tip is in.
-            // If it's in epoch 2.x, then we must always begin a new tenure, but we can't do so
-            // right now since this sortition has no winner.
-            let (cur_stacks_tip_ch, _cur_stacks_tip_bh) =
-                SortitionDB::get_canonical_stacks_chain_tip_hash(self.sortdb.conn())
-                    .expect("FATAL: failed to query sortition DB for stacks tip");
-
-            let stacks_tip_sn =
-                SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), &cur_stacks_tip_ch)
-                    .expect("FATAL: failed to query sortiiton DB for epoch")
-                    .expect("FATAL: no sortition for canonical stacks tip");
-
-            let cur_epoch =
-                SortitionDB::get_stacks_epoch(self.sortdb.conn(), stacks_tip_sn.block_height)
-                    .expect("FATAL: failed to query sortition DB for epoch")
-                    .expect("FATAL: no epoch defined for existing sortition");
-
-            if cur_epoch.epoch_id < StacksEpochId::Epoch30 {
-                debug!(
-                    "As of sortition {}, there has not yet been a Nakamoto tip. Cannot mine.",
+        self.tenure_extend_timeout = None;
+
+        if sn.sortition {
+            // a sortition happened
+            if won_sortition || self.config.get_node_config(false).mock_mining {
+                // a sortition happenend, and we won
+                info!("Relayer: Won sortition; begin tenure.";
+                      "winning_sortition" => %sn.consensus_hash);
+                return Some(MinerDirective::BeginTenure {
+                    parent_tenure_start: committed_index_hash,
+                    burnchain_tip: sn.clone(),
+                    election_block: sn,
+                    late: false,
+                });
+            }
+
+            let mining_pkh_opt = self.get_mining_key_pkh();
+
+            // a sortition happened, but we didn't win.
+            match Self::can_continue_tenure(
+                &self.sortdb,
+                &mut self.chainstate,
+                sn.consensus_hash,
+                mining_pkh_opt,
+            ) {
```

Comment:
Does `can_continue_tenure` need to be invoked here? After the recent changes to this PR, `can_continue_tenure` checks if there's a sortition in the most recent snapshot, and if there is, it returns `None`. I think this can just be replaced with `return Some(MinerDirective::StopTenure)` -- everything else is functionally dead code.

Reply:
Yes, I can remove this and just stop the tenure.

Reply:
I think this was a bad suggestion -- `can_continue_tenure` _does_ return `Ok(Some(..))` here in the event that the next tenure is a reorg, which is the correct behavior (and its roughly how the miner works in develop already). That's why `miner_forking` and `tenure_extend_after_bad_commit` started failing.

Reply:
D'oh! Should have seen the `self.tenure_extend_timeout = Some(Instant::now());` as a side-effect that would have been removed.